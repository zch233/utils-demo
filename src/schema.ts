export const FIELD_TYPES = [
    {
        label: '文本',
        value: 'string',
    },
    {
        label: '数字',
        value: 'number',
    },
    {
        label: '布尔',
        value: 'boolean',
    },
    {
        label: '对象',
        value: 'object',
    },
    {
        label: '数组',
        value: 'array',
    },
] as const;

export type SchemaType = (typeof FIELD_TYPES)[number]['value'];
export interface FieldItem {
    key: string;
    title: string;
    type: SchemaType;
    required: 0 | 1;
    description: string;
    children?: FieldItem[];
    sortedProps: string[];
    usedByScrollIntoView?: 1;
}
export interface JsonSchema {
    type: SchemaType;
    title: string;
    description: string;
    properties?: { [key: string]: JsonSchema };
    items?: JsonSchema;
    required?: string[];
    sortedProps: string[];
}

/**
 * 生成器的表单项 转为 schema 格式的数据
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/schema.html#form2Schema
 */
export function form2Schema(input: FieldItem): JsonSchema {
    const schema: JsonSchema = { type: input.type, title: input.title, description: input.description, sortedProps: [] };

    if (input.type === 'object') {
        schema.properties = {};
        schema.required = [];
        schema.sortedProps = input.sortedProps;
        if (input.sortedProps && input.sortedProps.length > 0) {
            input.sortedProps.forEach(key => {
                if (!input.children || input.children.length === 0) return;
                const child = input.children.find(c => c.key === key);
                if (!child) return;
                if (child.required) {
                    schema.required?.push(child.key);
                }
                schema.properties![child.key] = form2Schema(child);
            });
        }
    } else if (input.type === 'array' && input.children && input.children.length > 0) {
        schema.items = form2Schema(input.children[0]);
    }

    return schema;
}

/**
 * schema 格式的数据 转为 生成器的表单项
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/schema.html#schema2form
 */
export function schema2Form(schema: JsonSchema): FieldItem {
    const input: FieldItem = {
        type: schema.type,
        title: schema.title,
        description: schema.description,
        key: schema.items ? 'items' : 'root',
        required: 1,
        sortedProps: schema.sortedProps,
    };

    if (schema.type === 'object') {
        input.children = [];
        if (schema.properties) {
            if (!schema.sortedProps?.length) {
                schema.sortedProps = Object.keys(schema.properties) as string[];
            }
            schema.sortedProps?.forEach(key => {
                const childSchema = schema.properties![key];
                const childItem = schema2Form(childSchema);
                childItem.key = key;
                if (schema.required && schema.required.includes(key)) {
                    childItem.required = 1;
                } else {
                    childItem.required = 0;
                }
                input.children!.push(childItem);
            });
        }
    } else if (schema.type === 'array' && schema.items) {
        const childItem = schema2Form(schema.items);
        input.children = [childItem];
    }

    return input;
}
