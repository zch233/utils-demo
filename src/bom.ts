import queryString from 'query-string';

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/bom.html#getparams
 */
export const getParams = (key?: string, href?: string | URL, isHash?: boolean) => {
    const realURL = new URL(href || window.location.href);
    const params = queryString.parse(isHash ? realURL.hash.replace(/.*(?=\?)/, '') : realURL.search);
    return key ? params[key] : params;
};

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/bom.html#geturlparams
 */
export const getURLParams = (key?: string, href?: string | URL) => getParams(key, href);

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/bom.html#geturlhashparams
 */
export const getURLHashParams = (key?: string, href?: string | URL) => getParams(key, href, true);

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/bom.html#getbrowsers
 */
export const getBrowsers = () => {
    const u = navigator.userAgent;
    return {
        //移动终端浏览器版本信息
        trident: u.includes('Trident'), // IE内核
        presto: u.includes('Presto'), // opera内核
        webKit: u.includes('AppleWebKit'), // 苹果、谷歌内核
        gecko: u.includes('Gecko') && !u.includes('KHTML'), // 火狐内核
        mobile: !!/AppleWebKit.*Mobile.*/.test(u), // 是否为移动终端
        ios: !!/\(i[^;]+;( U;)? CPU.+Mac OS X/.test(u), // ios终端
        android: u.includes('Android') || u.includes('Linux'), // android终端或者uc浏览器
        iPhone: u.includes('iPhone'), // 是否为iPhone或者QQHD浏览器
        iPad: u.includes('iPad'), // 是否iPad
        dingTalk: u.includes('DingTalk'), // 是否为钉钉内部
        alipay: u.includes('AlipayClient'), // 是否为支付宝
        alipayMini: u.toLowerCase().includes('miniprogram') && u.toLowerCase().includes('alipay'), // 支付宝小程序
        wechat: /micromessenger/i.test(u), // 是否微信中
        wechatMini: u.toLowerCase().includes('miniprogram/wx') || (<any>window).__wxjs_environment === 'miniprogram', // 是否微信小程序中
        webApp: !u.includes('Safari'), // 是否web应该程序，没有头部与底部
        dtDreamApp: u.toLowerCase().includes('dtdreamweb'), // 浙里办 APP
        taurusApp: u.includes('TaurusApp'), // 是否为专有钉钉和浙政钉钉
        ykbapp: u.toLowerCase().includes('ykbapp'), // 是否愉快办 APP
    };
};

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/bom.html#isInIframe
 */
export const isInIframe = () => window.frames.length !== window.parent.frames.length;

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/bom.html#isInMicroApp
 */
// @ts-expect-error window __MICRO_APP_ENVIRONMENT__
export const isInMicroApp = (): boolean => window.__MICRO_APP_ENVIRONMENT__;
