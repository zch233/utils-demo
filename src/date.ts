import dayjs from 'dayjs';

// https://release.group-ds.com/dev-newbee-handbook/utils/utils.html#getMonthWeekCount
export const getMonthWeekCount = (month?: string, fromStart = 1) => {
    // fromStart -> 定义周的第一天是周几,默认是周一
    // w -> 计算该月1号是周几,0是周日.
    const w = dayjs(month).day();
    // 该月的天数
    const d = dayjs(month).daysInMonth();
    // 根绝fromStart和w的结果,算出第一周有几天
    const firstWeekDays = w ? (7 + fromStart - w) % 7 : fromStart ? fromStart : 7 - fromStart;
    // 计算该月有几周
    return Math.ceil((d - firstWeekDays) / 7) + 1;
};

// https://release.group-ds.com/dev-newbee-handbook/utils/utils.html#getMonthWeeks
export const getMonthWeeks = (
    month?: string
): {
    key: string;
    start_time: string;
    end_time: string;
    active: boolean;
}[] => {
    const m = dayjs(month).format('YYYY-MM');
    const monthWeekCount = getMonthWeekCount(m);
    const monthFirstDayReal = dayjs(m).startOf('month');
    // dayjs默认一周是以周日开始，所以如果该月的第一天是周日，我们提前减一天再计算
    const monthFirstDay = monthFirstDayReal.day() === 0 ? monthFirstDayReal.add(-1, 'days') : monthFirstDayReal;
    const monthLastDay = dayjs(m).endOf('month');
    return [...Array(monthWeekCount).keys()].map(index => {
        const startTime = index === 0 ? monthFirstDayReal : dayjs(monthFirstDay).add(index, 'week').startOf('week').add(1, 'days');
        const endTime = index === monthWeekCount - 1 ? monthLastDay : dayjs(monthFirstDay).add(index, 'week').endOf('week').add(1, 'days');
        return {
            key: `${startTime.format('YYYY-MM-DD')}-${endTime.format('YYYY-MM-DD')}`,
            start_time: startTime.format('YYYY-MM-DD'),
            end_time: endTime.format('YYYY-MM-DD'),
            active: dayjs().isAfter(startTime) && dayjs().isBefore(endTime),
        };
    });
};
