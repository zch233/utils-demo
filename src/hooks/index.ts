export * from './useCountDown';
export * from './useEventListener';
export * from './useKeepScroll';
export * from './useLazyRender';
export * from './useLockScroll';
export * from './useMountedOrActivated';
export * from './usePageVisibility';
export * from './useRaf';
export * from './useRect';
export * from './useRoute';
export * from './useScrollParent';
export * from './useScrollTo';
export * from './useSetKeepAlive';
export * from './useTouch';
export * from './useVisibilityChange';
export * from './useWindowSize';
