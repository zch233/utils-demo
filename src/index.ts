export * from 'lodash-unified';
export * from './utils';
export * from './bom';
export * from './screen';
export * from './validate';
export * from './hooks';
export * from './schema';
export * from './vendor';
