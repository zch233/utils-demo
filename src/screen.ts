/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/screen.html#parsehexcolor
 */
export const parseHexColor = function (hexStr: string) {
    return hexStr.length === 4
        ? hexStr
              .slice(1)
              .split('')
              .map(function (s) {
                  return 0x11 * Number.parseInt(s, 16);
              })
        : [hexStr.slice(1, 3), hexStr.slice(3, 5), hexStr.slice(5, 7)].map(function (s) {
              return Number.parseInt(s, 16);
          });
};

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/screen.html#gradientcolors
 */
export const gradientColors = (start: string, end: string, steps: number, gamma: number) => {
    let i, j, ms, me;
    const output: any = [];
    const so: any = [];
    gamma = gamma || 1;
    const normalize = function (channel: number) {
        return (channel / 255) ** gamma;
    };
    const pad = function (s: string) {
        return s.length === 1 ? `0${s}` : s;
    };
    const _start = parseHexColor(start).map(normalize);
    const _end = parseHexColor(end).map(normalize);
    for (i = 0; i < steps; i++) {
        ms = i / (steps - 1);
        me = 1 - ms;
        for (j = 0; j < 3; j++) {
            so[j] = pad(Math.round((_start[j] * me + _end[j] * ms) ** (1 / gamma) * 255).toString(16));
        }
        output.push(`#${so.join('')}`);
    }
    return output;
};

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/screen.html#geojson
 */
interface geoJson {
    coordinates: Array<any>;
    lnglat: Array<string | number>;
    code: number | string;
    name: string;
    label: string;
    value: number | string;
    level: number;
    children?: any[];
}
export const createGeoJson = (fetchData: Array<geoJson>) => {
    return {
        type: 'FeatureCollection',
        features: fetchData.map(o => {
            const { coordinates, lnglat, code, name, level, label, value, children } = o;
            return {
                type: 'Feature',
                properties: {
                    code: code || value,
                    name: name || label,
                    cp: lnglat,
                    level,
                    children: children || [],
                },
                geometry: {
                    type: 'MultiPolygon',
                    coordinates,
                },
            };
        }),
    };
};
