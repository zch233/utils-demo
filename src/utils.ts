import { set } from 'lodash-unified';
import type { AxiosResponse } from 'axios';

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/utils.html#base64tofile
 */
export const base64ToFile = (base64: string, filename?: string) => {
    const arr = base64.split(',');
    const mime = arr[0].match(/:(.*?);/)?.[1];
    const bstr = atob(arr[1]);
    let n = bstr.length;
    const u8arr = new Uint8Array(n);
    while (n--) {
        u8arr[n] = bstr.charCodeAt(n);
    }
    return new File([u8arr], filename || Date.now().toString(), { type: mime });
};

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/utils.html#blobtofile
 */
export const blobToFile = (blob: Blob, filename?: string, options?: FilePropertyBag) =>
    new File([blob], filename || Date.now().toString(), { type: blob.type, ...options });

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/utils.html#openlinkviaanchor
 */
export const openLinkViaAnchor = (url: string, anchorAttrs?: Record<string, any>) => {
    const anchor = document.createElement('a');
    anchor.href = url;
    if (anchorAttrs) Object.keys(anchorAttrs).map(key => set(anchor, key, anchorAttrs[key]));
    anchor.click();
    anchor.remove();
};

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/utils.html#openlink
 */
export const openLink = (url: string, target?: Record<string, any> | string, windowFeatures?: string) => {
    const windowOpen = window.open(url, typeof target === 'string' ? target : undefined, windowFeatures);
    if (windowOpen == null || typeof windowOpen === 'undefined') {
        openLinkViaAnchor(url, typeof target === 'string' ? { target } : target);
    }
};

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/utils.html#downloadwithblob
 */
interface DownloadWithBlob {
    (_response: AxiosResponse, filename?: string): void;
    (_response: Blob, filename: string): void;
}
export const downloadWithBlob: DownloadWithBlob = (_response, filename) => {
    const blobData = (_response as AxiosResponse).data || _response;
    return new Promise((resolve, reject) => {
        if (blobData instanceof Blob) {
            const fileReader = new FileReader();
            fileReader.addEventListener('load', function () {
                const blob = blobData;
                // 本地保存文件
                const url = window.URL.createObjectURL(blob) || window.webkitURL?.createObjectURL(blob);
                // TODO：注意这里的 headers 用于获取文件名称，确保后台已经返回正确字段
                const _filename = (_response as AxiosResponse).headers?.['content-disposition']?.split('filename*=')?.[1]?.slice(7);
                openLinkViaAnchor(url, { download: decodeURIComponent(filename || _filename) });
                window.URL.revokeObjectURL(url);
                resolve(blobData);
            });
            fileReader.readAsText(blobData);
        } else {
            console.error('downloadWithBlob：非文件流响应');
            reject(_response);
        }
    });
};

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/utils.html#stringtohex
 */
export const stringToHex = (str: string) => str.split('').reduce((res, v, i) => res + str.charCodeAt(i).toString(16), '');

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/utils.html#stringtobin
 */
export const stringToBin = (str: string) =>
    String.fromCharCode.apply(
        String,
        str
            .split(/(.{2})/)
            .filter(Boolean)
            .reduce((res, v) => res.concat(Number.parseInt(v, 16)), [] as number[])
    );

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/utils.html#numbertochinese
 */
export const numberToChinese = (str: number | string): string => {
    let num: number = Number(str);
    const chineseNums: string[] = ['零', '一', '二', '三', '四', '五', '六', '七', '八', '九'];
    const unitChars: string[] = ['', '十', '百', '千', '万'];

    if (num === 0 || Number.isNaN(num)) {
        return chineseNums[0];
    }

    let result: string = '';
    let unitIndex: number = 0;
    let needZero: boolean = false;

    while (num > 0) {
        const digit: number = num % 10;
        if (digit !== 0) {
            result = chineseNums[digit] + unitChars[unitIndex] + result;
            needZero = false;
        } else {
            if (!needZero) {
                result = chineseNums[digit] + result;
                needZero = true;
            }
        }

        num = Math.floor(num / 10);
        unitIndex++;
    }

    // 处理特殊情况：如果结果以"一十"开头，去掉前面的"一"
    result = result.startsWith('一十') ? result.substring(1) : result;
    // 处理特殊情况：如果结果以"零"结尾，去掉后面的"零"
    result = result.endsWith('零') ? result.substring(0, result.length - 1) : result;

    return result;
};

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/utils.html#measuretext
 */
export class MeasureText {
    private canvas: HTMLCanvasElement;
    private context: CanvasRenderingContext2D;

    constructor(font: string = '14px Microsoft YaHei') {
        this.canvas = document.createElement('canvas');
        this.context = this.canvas.getContext('2d')!;
        this.context.font = font;
    }

    getWidth(text: string): number {
        const metrics = this.context.measureText(text);
        return metrics.width;
    }
}

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/utils.html#array2keyobject
 */
interface Item {
    [key: string]: any;
}

interface KeyedObject<T> {
    [key: string]: T;
}
export const array2KeyObject = <T extends Item>(originalArray: T[] = [], key: keyof T = 'code', value?: keyof T): KeyedObject<T> | KeyedObject<any> => {
    return originalArray.reduce((acc: KeyedObject<T> | KeyedObject<any>, item: T) => {
        if (!item || !item[key]) return acc;
        if (value) {
            acc[item[key] as string] = item[value];
        } else {
            acc[item[key] as string] = item;
        }
        return acc;
    }, {});
};

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/utils.html#btoa2
 */
export const btoa2 = (text: string) => btoa(String.fromCharCode(...new TextEncoder().encode(text)));

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/utils.html#atob2
 */
export const atob2 = (text: string) => new TextDecoder().decode(Uint8Array.from(atob(text), c => c.charCodeAt(0)));

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/utils.html#desensitizationphone
 */
export const desensitizationPhone = (phone: string) => phone && phone.toString().replace(/(.{3}).*(.{4})/, '$1***$2');

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/utils.html#desensitizationidcard
 */
export const desensitizationIDCard = (cardNo: string) => {
    if (!cardNo) return cardNo;
    const prefix = cardNo.slice(0, 6); // 获取前三位身份证号
    const suffix = cardNo.slice(-4); // 获取后四位身份证号
    // 生成中间部分的脱敏字符，长度与原身份证号中间部分相同
    const mask = '*'.repeat(cardNo.length - 10);
    return `${prefix}${mask}${suffix}`;
};

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/utils.html#desensitizationname
 */
export const desensitizationName = (name: string) => {
    if (name.length <= 1 || !name) {
        return name;
    }
    if (name.length === 2) {
        return `${name.slice(0, 1)}*`; // 如果字符串长度为2，返回第一个字符加一个星号
    }
    const firstChar = name.slice(0, 1); // 获取首字符
    const lastChar = name.slice(-1); // 获取尾字符
    const maskedLength = name.length - 2; // 计算中间部分长度
    return `${firstChar}${'*'.repeat(maskedLength)}${lastChar}`;
};

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/utils.html#desensitizationinfo
 */
const isNumberString = (str: string): boolean => /^\d+$/.test(str.trim());

export const desensitizationInfo = (str: string) => {
    const strLength = str?.length ?? 0;
    if (!str) return str;
    if (strLength === 11 && isNumberString(str)) {
        return desensitizationPhone(str);
    } else if ((strLength === 15 && isNumberString(str)) || (strLength === 18 && (isNumberString(str) || str.includes('X')))) {
        return desensitizationIDCard(str);
    } else {
        return desensitizationName(str);
    }
};
/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/utils.html#getCssVariable
 */
export const getCssVariable = (variable = '--color-master'): string => String(getComputedStyle(document.documentElement).getPropertyValue(variable)).trim();

/**
 * 获取原始类型
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/utils.html#getPrimitiveType
 */
type PrimitiveType =
    | 'undefined'
    | 'null'
    | 'boolean'
    | 'number'
    | 'string'
    | 'symbol'
    | 'bigint'
    | 'object'
    | 'array'
    | 'function'
    | 'date'
    | 'regexp'
    | 'error'
    | 'map'
    | 'set'
    | 'weakmap'
    | 'weakset'
    | 'arraybuffer'
    | 'dataview'
    | 'promise'
    | 'generator'
    | 'generatorfunction'
    | 'asyncfunction'
    | 'proxy';
export const getPrimitiveType = (value: any): PrimitiveType =>
    /\[object (\w+)]/.exec(Object.prototype.toString.call(value))![1].toLocaleLowerCase() as PrimitiveType;
