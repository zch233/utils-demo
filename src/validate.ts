/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/validate.html#validateidcard
 */
export const validateIDCard = (idCard: string) => {
    idCard = idCard.toString();
    const city: Record<number, string> = {
        11: '北京',
        12: '天津',
        13: '河北',
        14: '山西',
        15: '内蒙古',
        21: '辽宁',
        22: '吉林',
        23: '黑龙江',
        31: '上海',
        32: '江苏',
        33: '浙江',
        34: '安徽',
        35: '福建',
        36: '江西',
        37: '山东',
        41: '河南',
        42: '湖北',
        43: '湖南',
        44: '广东',
        45: '广西',
        46: '海南',
        50: '重庆',
        51: '四川',
        52: '贵州',
        53: '云南',
        54: '西藏',
        61: '陕西',
        62: '甘肃',
        63: '青海',
        64: '宁夏',
        65: '新疆',
        71: '台湾',
        81: '香港',
        82: '澳门',
        91: '国外',
    };
    let num = idCard.toUpperCase();
    if (!/(^\d{15}$)|(^\d{17}(\d|X)$)/.test(num)) {
        // throw '身份证位数错误！';
        return false;
    }
    if (city[Number.parseInt(num.substr(0, 2))] === undefined) {
        // throw '身份证格式错误！';
        return false;
    }
    const len = num.length;
    let re;
    if (len === 15) {
        re = /^(\d{6})(\d{2})(\d{2})(\d{2})(\d{3})$/;
        const arrSplit = num.match(re);
        const dtmBirth = new Date(`19${arrSplit?.[2]}/${arrSplit?.[3]}/${arrSplit?.[4]}`);
        const bGoodDay =
            // @ts-expect-error xxx
            dtmBirth.getYear() === Number(arrSplit?.[2]) && dtmBirth.getMonth() + 1 === Number(arrSplit?.[3]) && dtmBirth.getDate() === Number(arrSplit?.[4]);
        if (!bGoodDay) {
            // throw '身份证日期格式错误';
            return false;
        } else {
            const arrInt = [7, 9, 10, 5, 8, 4, 2, 1, 6, 3, 7, 9, 10, 5, 8, 4, 2];
            const arrCh = ['1', '0', 'X', '9', '8', '7', '6', '5', '4', '3', '2'];
            let nTemp = 0;
            num = `${num.substr(0, 6)}19${num.substr(6, num.length - 6)}`;
            for (let i = 0; i < 14; i++) {
                nTemp += Number(num.substr(i, 1)) * arrInt[i];
            }
            num += arrCh[nTemp % 11];
            return true;
        }
    }
    if (len === 18) {
        re = /^(\d{6})(\d{4})(\d{2})(\d{2})(\d{3})(\d|X)$/;
        const arrSplit = num.match(re);
        const dtmBirth = new Date(`${arrSplit?.[2]}/${arrSplit?.[3]}/${arrSplit?.[4]}`);
        const bGoodDay =
            dtmBirth.getFullYear() === Number(arrSplit?.[2]) &&
            dtmBirth.getMonth() + 1 === Number(arrSplit?.[3]) &&
            dtmBirth.getDate() === Number(arrSplit?.[4]);
        if (!bGoodDay) {
            // throw '身份证日期格式错误';
            return false;
        } else {
            const arrInt = [7, 9, 10, 5, 8, 4, 2, 1, 6, 3, 7, 9, 10, 5, 8, 4, 2];
            const arrCh = ['1', '0', 'X', '9', '8', '7', '6', '5', '4', '3', '2'];
            let nTemp = 0;
            for (let i = 0; i < 17; i++) {
                nTemp += Number(num.substr(i, 1)) * arrInt[i];
            }
            const valnum = arrCh[nTemp % 11];
            if (valnum !== num.substr(17, 1)) {
                // throw '身份证格式错误！';
                return false;
            }
            return true;
        }
    }
};

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/validate.html#validatephone
 */
export const validatePhone = (phone: string) => /^(0\d{2,3}-?\d{7,8}|0\d{9,11}|1[3-9]\d{9})$/.test(phone);

/**
 * 文档地址：https://release.group-ds.com/dev-newbee-handbook/utils/validate.html#validatename
 */
export const validateName = (name: string) => /^[\u3400-\u4DB5\u4E00-\u9FCB]{2,10}([\u00B7\u25CF][\u3400-\u4DB5\u4E00-\u9FCB]{2,10})*$/gi.test(name);
