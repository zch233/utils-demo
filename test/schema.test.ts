import { describe, expect, it } from 'vitest';
import { form2Schema, schema2Form } from '../src';

// 使用 vitest 帮我测试这两个方法是否正确
describe('form2Schema and schema2Form should be correct', () => {
    it('schema2Form should be correct', () => {
        const schema = schema2Form({
            type: 'object',
            title: '对象',
            sortedProps: ['a', 'b'],
            description: '对象描述',
            required: ['a'],
            properties: {
                a: {
                    type: 'string',
                    title: 'a',
                    description: 'a',
                    sortedProps: ['a'],
                },
                b: {
                    type: 'string',
                    title: 'b',
                    description: 'b',
                    sortedProps: ['b'],
                },
            },
        });
        expect(schema).toMatchInlineSnapshot(`
          {
            "children": [
              {
                "description": "a",
                "key": "a",
                "required": 1,
                "sortedProps": [
                  "a",
                ],
                "title": "a",
                "type": "string",
              },
              {
                "description": "b",
                "key": "b",
                "required": 0,
                "sortedProps": [
                  "b",
                ],
                "title": "b",
                "type": "string",
              },
            ],
            "description": "对象描述",
            "key": "root",
            "required": 1,
            "sortedProps": [
              "a",
              "b",
            ],
            "title": "对象",
            "type": "object",
          }
        `);
    });

    it('form2Schema should be correct', () => {
        const schema = form2Schema({
            children: [
                {
                    description: 'a',
                    key: 'a',
                    required: 1,
                    sortedProps: ['a'],
                    title: 'a',
                    type: 'string',
                },
                {
                    description: 'b',
                    key: 'b',
                    required: 0,
                    sortedProps: ['b'],
                    title: 'b',
                    type: 'string',
                },
            ],
            description: '对象描述',
            key: 'root',
            required: 1,
            sortedProps: ['a', 'b'],
            title: '对象',
            type: 'object',
        });
        expect(schema).toMatchInlineSnapshot(`
          {
            "description": "对象描述",
            "properties": {
              "a": {
                "description": "a",
                "sortedProps": [],
                "title": "a",
                "type": "string",
              },
              "b": {
                "description": "b",
                "sortedProps": [],
                "title": "b",
                "type": "string",
              },
            },
            "required": [
              "a",
            ],
            "sortedProps": [
              "a",
              "b",
            ],
            "title": "对象",
            "type": "object",
          }
        `);
    });
});
