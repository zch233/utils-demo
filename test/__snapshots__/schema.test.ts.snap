// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`form2Schema and schema2Form > form2Schema and schema2Form should be correct 1`] = `
{
  "children": [
    {
      "description": "a",
      "key": "a",
      "required": 0,
      "sortedProps": [
        "a",
      ],
      "title": "a",
      "type": "string",
    },
    {
      "description": "b",
      "key": "b",
      "required": 0,
      "sortedProps": [
        "b",
      ],
      "title": "b",
      "type": "string",
    },
  ],
  "description": "对象描述",
  "key": "root",
  "required": 1,
  "sortedProps": [
    "a",
    "b",
  ],
  "title": "对象",
  "type": "object",
}
`;
